### PR TITLE
Exercise 5

### PR DESCRIPTION
**Pull request by KarakoA**
We have given 2 reviews:

- Systems-Development-and-Frameworks/lichtow#11 (review) reviewed by @jingyophuong
- https://github.com/Systems-Development-and-Frameworks/raum9/pull/6 reviewed by @kipmann

### CODE REVIEW

**Review by kipmann — COMMENTED** [review]
no remarks found

**Review by jingyophuong — COMMENTED** [review]
I think our work on code is already complete. Nice 👍👍👍👍👍👍👍

File: backend/src/resolvers/mutate_posts.spec.js [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+    it("responds with created post", async () => {
+      await expect(actionWritePost()).resolves.toMatchObject({
+        errors: undefined,
+        data: {
+          write: {
+            title: "Some post",
+            id: expect.any(String),
+            votes: 0,
+            author: { name: "Peter" },
```
Comment (medizinmensch) [inline backend/src/resolvers/mutate_posts.spec.js:62]: Here, you're not following the Arrange-Act-Assert principle. You Arrange in your beforeAll and Assert here, but you Act in another test. This is not desired. 

This is just one example.

**Review by medizinmensch — APPROVED** [review]
Hi Team!

I liked your in depth description on how to setup your app. 
I would recommend adding `nodemon` to your `package.json` instead of relying on the user to have it installed globally. 

Sadly, I had issues with your tests. I see that you turned on `--runInBand` which makes the test sequential. This already indicates that they depend on each other. Once I turned that off I got:

```
Test Suites: 3 failed, 1 passed, 4 total
Tests:       7 failed, 17 passed, 24 total
```

If I only run `mutations › WRITE_POST › responds with created post`, I also get an error. 
If you look at the code of your test cases, you see that they depend on each other and have side effects at least in two places. So I have to deduct 2 ⭐  for that :x: 

Also, I can't see which group you gave a review 🤔 We asked to put it in the description of the post. :x:

Apart from that:

:star: For choosing a scenario   and writing installation instructions in README.md.
:star:   For explaining WHY you have chosen the scenario in the README.md.
:star: For not committing   sensitive secrets (e.g. API keys) unencrypted to the repository.
:star:   :star: Data created in mutations are persisted and your queries and mutations   still work as expected.
:star:   Your database is never left in an invalid state.
:star: You deny the client to   call mutations and queries of your subschema directly.


So all in all that makes 7/10 :star: 

Gif

**Comment by KarakoA** [comment]
> Hi Team!
> 
> I liked your in depth description on how to setup your app.
> I would recommend adding `nodemon` to your `package.json` instead of relying on the user to have it installed globally.
> 
> Sadly, I had issues with your tests. I see that you turned on `--runInBand` which makes the test sequential. This already indicates that they depend on each other. Once I turned that off I got:
> 
> ```
> Test Suites: 3 failed, 1 passed, 4 total
> Tests:       7 failed, 17 passed, 24 total
> ```
> 
> If I only run `mutations › WRITE_POST › responds with created post`, I also get an error.
> If you look at the code of your test cases, you see that they depend on each other and have side effects at least in two places. So I have to deduct 2 ⭐ for that ❌
> 
> Also, I can't see which group you gave a review 🤔 We asked to put it in the description of the post. ❌
> 
> Apart from that:
> 
> ⭐ For choosing a scenario and writing installation instructions in README.md.
> ⭐ For explaining WHY you have chosen the scenario in the README.md.
> ⭐ For not committing sensitive secrets (e.g. API keys) unencrypted to the repository.
> ⭐ ⭐ Data created in mutations are persisted and your queries and mutations still work as expected.
> ⭐ Your database is never left in an invalid state.
> ⭐ You deny the client to call mutations and queries of your subschema directly.
> 
> So all in all that makes 7/10 ⭐
> 
> Gif

Hey Youri!
Thanks a lot for the review, we were planning to do a review for @Systems-Development-and-Frameworks/raum9, as always, and are waiting for them to be ready with their PR. 

It was my mistake of assigning @Systems-Development-and-Frameworks/mentors  to the PR, I assumed you review after the deadline and we still have time for everything until then. We will review @Systems-Development-and-Frameworks/raum9's PR as soon as it is done (still before Wednesday). 

It would be awesome if we can get the point for that, once we've done it. 

I don't agree with the deduction of 2 points for side effect free testsing.
From the task description of [Exercise5](https://github.com/Systems-Development-and-Frameworks/homework/tree/main/exercises/5):
>Ensure that your software tests have no side-effects, ie. your tests do not interfere with each other. Any software test case that introduces a side-effect will cost you a ⭐. Do not simply delete software tests that were required previously. You can ensure side-effect-free tests by either:

>Cleaning the database before or after your tests 
 
That's exactly what we did, before and after each tests the database is cleaned and populated if needed. Of course they have to run `--inBand`, how else should this be accomplished without mocking a database/having a separate database for each test?

I ran
```bash
npm run test --  "mutate_posts.spec.js" -t "mutations WRITE_POST responds with created post"
```
and it succeeds. I don't see why it shouldn't. The database is cleared and some data is populated after in the scoped `beforeEach` of the test file

File: backend/.env.test [hunk 2]
```diff
@@ -1 +1,7 @@
 JWT_SECRET=SECRET_1234
+NEO4J_USERNAME=NEO4J
```
[the remark this answers is not in the document]
Comment (KarakoA) [reply backend/.env.test:2]: Sorry for that! I had authentication off in my local db setup and it slipped through
```
# Whether requests to Neo4j are authenticated.
# To disable authentication, uncomment this line
dbms.security.auth_enabled=false
```

File: README.md [hunk 3]
```diff
@@ -27,3 +27,69 @@ Save: update the todo and display it
 Cancel: cancel the form submission
 
 Delete: delete the todo
+
+## Homework - Exercise #5
+### Chosen Scenario - Neo4J and neo4j-graphql-js
+
+#### Motivation
+
+Why did we choose Neo4J and neo4j-graphql-js? Lyon's presentation convinced us to use and learn about this interesting technology. A similar use case like our WebApp was explained very well in the lecture and we would like to go deeper into this technology to understand how our application can look like using the graphs. It is also very interesting to see how the elements of the application depend on each other. This technology is new for all members, so we would like to explore it further. Regarding our use case: Our data can be represented very clearly with Neo4J, at the same time it also can be visualised and explored very well with Neo4J.
```
Comment (JannisBe) [inline README.md:36]: i like the motivation and i think its sufficient

File: backend/package.json [hunk 4]
```diff
@@ -4,15 +4,19 @@
   "description": "",
   "main": "src/index.js",
   "scripts": {
-    "test": "jest --colors",
+    "test": "jest --colors --runInBand",
```
Comment (JannisBe) [inline backend/package.json:7]: this makes every test run sequentially, why is it needed? By running everything parallel you increase the chance of finding side effects, dont you?

File: backend/src/db/fixture.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+const brothersPassword = "brother1";
+
+const petersPost = new Post({
+  title: "Pinguine sind keine Vögel",
+  author: peter,
+  id: "613e8596-b432-4b74-89d4-06709644d8ee",
+});
+
+export default { peter, brother, petersPassword, brothersPassword, petersPost };
```
Comment (JannisBe) [inline backend/src/db/fixture.js:26]: although i like the approach of refactoring the dummy db, i think its better to put them into the test files directly. I guess the best approach is to use the factory Pattern and just create FactoryObjects here, but im not sure if graphQL gives the opportunity to do this

File: backend/src/permissions.js [hunk 6]
```diff
@@ -1,23 +1,35 @@
...
   }
 );
 
 export const permissions = shield(
   {
+    User: allow,
+    Post: allow,
+
     Query: {
-      "*": allow,
```
Comment (JannisBe) [inline backend/src/permissions.js:13]: `'*': deny,`
for UserData i like the explicit is better than implicit approach more, also you can find errors in your permissions more easily

**Comment by jingyophuong** [comment]
https://github.com/Systems-Development-and-Frameworks/lichtow/pull/11#pullrequestreview-561866347 reviewed by @jingyophuong

File: backend/src/resolvers/mutate_users.spec.js [hunk 7]
```diff
@@ -57,23 +57,23 @@ describe("mutations", () => {
...
+        "Geogre",
+        "geogre@widerstand-der-pinguine.ev",
+        "sfseltuowu3979fdk"
       );
       expect(errors).toBeUndefined();
       let verified = verifyToken(signup);
       expect(verified).toEqual({
-        exp: expect.anything(),
-        iat: expect.anything(),
+        exp: expect.any(Number),
```
Comment (JannisBe) [inline backend/src/resolvers/mutate_users.spec.js]: very thoroughly tested, well done

**Review by JannisBe — COMMENTED** [review]
In general i think its a nice PR and i like your way of testing everything very detailed. There were only minor issues i found regarding some security concepts, besides the things already mentioned by others. Also id suggest using a different approach for your test Database.

**Comment by medizinmensch** [comment]
Hi code_party_kp!

Last Wednesday, we talked about the last two points I did not yet give you. I somehow mixed up your review with another one (I thought you only had one `spec.js` file but you have multiple). Additionally, while I tried to explicitly start that one flaky test, I started two instead of one (because of naming similarities). 
So your test cases are good. Sorry for the confusion and thanks for clarification!
Since you also got your review now, you get additional 3 :star:.

This makes 10/10 ⭐ , congrats!


Nice

**Pull request merged**